numpy/core/src/multiarray/einsum.dispatch.c.src

```diff
@@ -0,0 +1,1866 @@
+        npyv_storel_f32(&accum, accum_sse);
+
+        /* Finish off the loop */
+        goto finish_after_unrolled_loop;
+    }
+#elif NPY_SIMD_F64 && @float64@
+    /* Use aligned instructions if possible */
+    if (EINSUM_IS_ALIGNED(data0) && EINSUM_IS_ALIGNED(data1)) {
+        /* Unroll the loop by UNROLL_BY */
+        while (count >= UNROLL_BY) {
+            count -= UNROLL_BY;
+            npyv_prefetch(data0 + 512, 1);
+            npyv_prefetch(data1 + 512, 1);
+
+/**begin repeat2
+ * #i = 0, 1, 2, 3#
+ */
+            /*
+             * NOTE: This accumulation changes the order, so will likely
+             *       produce slightly different results.
+             */
+            pack_size = @i@ * UNROLL_BY / 4;
+            a = npyv_mul_f64(npyv_loada_f64(data0+pack_size), npyv_loada_f64(data1+pack_size));
+            accum_sse = npyv_add_f64(accum_sse, a);
+/**end repeat2**/
+            data0 += UNROLL_BY;
+            data1 += UNROLL_BY;
+        }
+
+        /* Add the two SSE2 values and put in accum */
+        a = npyv_shuffle_f64(accum_sse, accum_sse, _MM_SHUFFLE2(0,1));
+        accum_sse = npyv_add_f64(a, accum_sse);
+        npyv_storel_f64(&accum, accum_sse);
+        
+        /* Finish off the loop */
+        goto finish_after_unrolled_loop;
+    }
+#endif
+
+    /* Unroll the loop by 8 */
+    while (count >= UNROLL_BY) {
+        count -= UNROLL_BY;
+
+#if NPY_SIMD && @float32@
+        npyv_prefetch(data0 + 512, 1);
+        npyv_prefetch(data1 + 512, 1);
+
+/**begin repeat2
+ * #i = 0, 1#
+ */
+        /*
+         * NOTE: This accumulation changes the order, so will likely
+         *       produce slightly different results.
+         */
+        pack_size = @i@ * UNROLL_BY / 2;
+        a = npyv_mul_f32(npyv_load_f32(data0+pack_size), npyv_load_f32(data1+pack_size));
+        accum_sse = npyv_add_f32(accum_sse, a);
+/**end repeat2**/
+#elif NPY_SIMD_F64 && @float64@
+        npyv_prefetch(data0 + 512, 1);
+        npyv_prefetch(data1 + 512, 1);
+
+/**begin repeat2
+ * #i = 0, 1, 2, 3#
+ */
+        /*
+         * NOTE: This accumulation changes the order, so will likely
+         *       produce slightly different results.
+         */
+        pack_size = @i@ * UNROLL_BY / 4;
+        a = npyv_mul_f64(npyv_load_f64(data0+pack_size), npyv_load_f64(data1+pack_size));
+        accum_sse = npyv_add_f64(accum_sse, a);
+/**end repeat2**/
+#else
+    for (int i = 0; i < UNROLL_BY; i++) {
+        accum += @from@(data0[i]) * @from@(data1[i]);
+    }
+#endif
+        data0 += UNROLL_BY;
+        data1 += UNROLL_BY;
+    }
+
+#if NPY_SIMD && @float32@
+    /* Add the four SSE values and put in accum */
+    a = npyv_shuffle_f32(accum_sse, accum_sse, _MM_SHUFFLE(2,3,0,1));
+    accum_sse = npyv_add_f32(a, accum_sse);
+    a = npyv_shuffle_f32(accum_sse, accum_sse, _MM_SHUFFLE(1,0,3,2));
+    accum_sse = npyv_add_f32(a, accum_sse);
+    npyv_storel_f32(&accum, accum_sse);
```

Comment:
Is the end result of this shuffle to convert `[a, b, c ,d]` to `[a + b + c + d]*4`?

Reply:
@eric-wieser you are right, every lane of vector `accum_sse `is the sum of all lanes.

Reply:
And then only one lane is actually used, right?

I think this would benefit from being extracted to a helper function, with an explanation of why such an odd approach is used.

Reply:
the odd approach has transformed to a more comprehended `sum `api.